proposals/lip-0045.md

```diff
@@ -984,164 +839,223 @@ The Interoperability module contains the following commands:
+6. `CCM_PROCESSED_STATUS_APPLIED`: The CCM was applied on the receiving chain.
+7. `CCM_PROCESSED_STATUS_DISCARDED`: The CCM was discarded.
+8. `CCM_PROCESSED_STATUS_RECOVERY_FORWARDED`: The CCM was forwarded to the original sending chain for recovery.
+9. `CCM_PROCESSED_STATUS_RECOVERY_APPLIED`: The CCM was recovered and applied on the original sending chain.
+10. `CCM_PROCESSED_STATUS_RECOVERY_DISCARDED`: The CCM was ricovered and discarded.
 
-This function has the following input parameters in the order given below:
 
-* `ccm`: The cross-chain message.
-* `errorStatus`: The error status to be set.
+##### Topics
 
-##### Returns
+- `sendingChainID`: The ID of the sending chain.
+- `receivingChainID`: The ID of the receiving chain.
+- `ccmID`: The ID of the CCM that was sent. This topic is present only for newly sent messages (event status equals `CCM_PROCESSED_STATUS_SENT`).
 
-A boolean indicating the success of the function execution.
+Furthermore, one
 
-##### Execution
+##### Data
+
+```java
+ccmProcessedDataSchema = {
+    "type": "object",
+    "required": ["ccmID", "status"],
+    "properties": {
+        "ccmID": {
+            "dataType": "bytes",
+            "length": HASH_LENGTH, 
+            "fieldNumber": 1
+        },
+        "status": {
+            "dataType": "uint32",
+            "fieldNumber": 2
+        }
+    }
+}
+```
+
+### Protocol Logic for Other Modules
+
+#### send
+
+The `send` function is used to create and add a message to the outbox of a partner chain.
 
 ```python
-error(ccm, errorStatus):
-    # Error codes from 0 to MAX_RESERVED_ERROR_STATUS (included) are reserved to the Interoperability module
-    if 0 <= errorStatus <= MAX_RESERVED_ERROR_STATUS:
-        return False
+def send(sendingAddress: Address, 
+    moduleID: ModuleID, 
+    crossChainCommandID: CrossChainCommandID, 
+    receivingChainID: ChainID, 
+    fee: uint64, 
+    status: uint32, 
+    params: object) -> None:
+    # Processing on the mainchain
+    if ownChainAccount.ID == MAINCHAIN_ID:
+        # Chain has to exist on mainchain
+        if chainAccount(receivingChainID) does not exists:
+            raise Exception('Receiving chain does not exist.')
+        partnerChainID = receivingChainID
+    # Processing on a sidechain
+    else:
+        # Check for direct channel
+        if chainAccount(receivingChainID) does not exists:
+            partnerChainID = mainchainID
+        else:
+            partnerChainID = receivingChainID
 
-    return sendInternal(
-        EMPTY_FEE_ADDRESS,
-        ccm.moduleID,
-        ccm.crossChainCommandID,
-        ccm.sendingChainID,
-        0,
-        errorStatus,
-        ccm.params
+    # Chain must be live; This checks is always on the receivingChainID
+    if not isLive(receivingChainID):
+        raise Exception('Receiving chain is not live.')
+    # Chain status must be active
+    if not chainAccount(partnerChainID).status == CHAIN_STATUS_ACTIVE:
```

Comment:
Should this check maybe come before checking liveness? For instance, for a terminated chain `isLive` already returns false, so doing this check first may result in a more specific exception case.

Reply:
chainAccount may not exist tho, so it would make for an extra check first

Reply:
Okay. I also see now that the liveness check is for the receiving chain and being active is checked for the partner chain.